### PR TITLE
nightfall-node: can sign and send txs from all routes

### LINKED ISSUE #1164: Implement optimist signing for all other proposer features
As a proposer, if I have an MVP, all other features can be signed by the optimist.

- [ ] Use the REST documentation, implement `send()` instead of the return of `encodeABI()` on the optimist

- [ ] Write unit tests for these function calls

Routes left

- [x] withdrawStake
- [x] withdraw > didnt have method in `nf3`, nor tests
- [x] payment > didnt have tests
- [x] encode > no `nf3` method, no updates beyond the endpoint for now

### PR DESCRIPTION
**Pull request by imagobea**
## What does this implement/fix? Explain your changes

- Optimist can sign and send txs directly from all routes. Updated: routes `withdrawStake`, `withdraw`, `payment`, `encode`

- Removed query param `ethAddress` from `pending-payments` and `stake` as the info is available from the node

- Updated `nf3` according to points above

- Removed sensitive logs from `app.mjs`

- Reviewed Swagger route `descriptions`. Reviewed swagger-options - some typos inconsistencies. Identified areas for future improvement.

- Tests: added ws closure to `circuits` and `administrator`. Further improvements to `proposer` - please note I have re-ordered some tests to improve the narrative of the file. There are still 4 skipped, hopefully to be updated in the upcoming PR.

- Upgraded `nightfall-sdk`

## Does this close any currently open issues?

Closes #1164 

## What commands can I run to test the change? 

You can look at the tests or try this with curl (remember most of the routes are now protected), eg

```
curl -k -X POST -H "X-APP-TOKEN: 422570cdb2b6a2cd805747a19592c21d620bf751dff0d62a35acebea63e2442d" -d "url=pepe&stake=999999999" http://localhost:8081/proposer/register
```

## Any other comments?

- With this PR all proposer/optimist endpoints are autonomous in the tx management

- This PR (and all the node PRs including the feature-branch) are already aligned with L2tokenisation

- Next I will work on a (final) PR to resolve Eth keys/address management, reduce code duplication, calculate tx gas, more test fixes

### CODE REVIEW

File: nightfall-optimist/src/routes/proposer.mjs [hunk 1]
```diff
@@ -52,21 +52,21 @@ router.post('/register', auth, async (req, res, next) => {
   const { url = '', stake = 0, fee = 0 } = req.body;
 
   try {
-    // Recreate Proposer, State contracts
-    const proposersContractInstance = await waitForContract(PROPOSERS_CONTRACT_NAME);
-    const proposersContractAddress = await getContractAddress(PROPOSERS_CONTRACT_NAME);
-    const stateContractInstance = await waitForContract(STATE_CONTRACT_NAME);
```
Comment (imagobea) [inline nightfall-optimist/src/routes/proposer.mjs:58]: Moved these after validations
(did this too later again in `/update` route)

File: nightfall-optimist/src/routes/proposer.mjs [hunk 2]
```diff
@@ -282,15 +298,12 @@ router.get('/pending-payments', auth, async (req, res, next) => {
 /**
  * Function to get stake for a proposer.
  */
-router.get('/stake', async (req, res, next) => {
-  const { proposerAddress } = req.query;
-  logger.debug(`requested stake for proposer ${proposerAddress}`);
+router.get('/stake', auth, async (req, res, next) => {
+  const ethAddress = req.app.get('ethAddress');
```
Comment (imagobea) [inline nightfall-optimist/src/routes/proposer.mjs:302]: Removed query params and protected route

File: nightfall-optimist/src/routes/block.mjs [hunk 3]
```diff
@@ -53,7 +53,7 @@ router.post('/check', auth, async (req, res, next) => {
  *      tags:
  *      - Block
  *      summary: Make block.
- *      description: Responsible to call the function that will generate a new block.
```
Comment (velkoskis) [inline nightfall-optimist/src/routes/block.mjs:56]: Thank you for these updates!

**Review by velkoskis — APPROVED** [review]
LGTM. Thank for adding WS closure on the tests and  making language updates too!
Feels good to wrap up nightfall-node routes.
Clean description and execution. Congrats!

File: nightfall-optimist/src/routes/proposer.mjs [hunk 4]
```diff
@@ -357,10 +349,26 @@ router.post('/de-register', auth, async (req, res, next) => {
...
   try {
-    const proposerContractInstance = await getContractInstance(PROPOSERS_CONTRACT_NAME);
-    const txDataToSign = await proposerContractInstance.methods.withdrawStake().encodeABI();
-    res.json({ txDataToSign });
+    // Recreate Proposer contract
+    const proposersContractInstance = await getContractInstance(PROPOSERS_CONTRACT_NAME);
+
+    // Withdraw proposer stake
+    const txDataToSign = await proposersContractInstance.methods.withdrawStake().encodeABI();
+    const tx = {
```
Comment (israelboudoux) [inline nightfall-optimist/src/routes/proposer.mjs:361]: I see the following block of code being repeated in this file, maybe we could create a method that does the job. Follows the repeated block:
```
    const tx = {
      from: ethAddress,
      to: proposersContractInstance.options.address,
      data: txDataToSign,
      gas: 8000000,
    };
    const signedTx = await web3.eth.accounts.signTransaction(tx, ethPrivateKey);
    const receipt = await web3.eth.sendSignedTransaction(signedTx.rawTransaction);
```

File: nightfall-optimist/src/routes/proposer.mjs [hunk 5]
```diff
@@ -503,11 +505,25 @@ router.get('/stake', auth, async (req, res, next) => {
...
+    const txDataToSign = await stateContractInstance.methods.withdraw().encodeABI();
+    const tx = {
+      from: ethAddress,
+      to: stateContractInstance.options.address,
+      data: txDataToSign,
+      gas: 8000000,
+    };
+    const signedTx = await web3.eth.accounts.signTransaction(tx, ethPrivateKey);
+    const receipt = await web3.eth.sendSignedTransaction(signedTx.rawTransaction);
+    logger.debug(`Transaction receipt ${receipt}`);
```
Comment (israelboudoux) [inline nightfall-optimist/src/routes/proposer.mjs:524]: Every object returned by the endpoints are automatically logged, then I think these `logger.debug` instructions that logs these objects can be removed.

File: nightfall-optimist/src/routes/proposer.mjs [hunk 6]
```diff
@@ -542,14 +561,33 @@ router.get('/withdraw', auth, async (req, res, next) => {
  */
 router.post('/payment', auth, async (req, res, next) => {
   const { blockHash } = req.body;
+  const ethAddress = req.app.get('ethAddress');
```
Comment (israelboudoux) [inline nightfall-optimist/src/routes/proposer.mjs:564]: Is this way of getting the value better than the one above? maybe we could use the best approach everywhere just to be uniform.

**Review by israelboudoux — APPROVED** [review]
LGTM, just a few things to work on I believe.

**Pull request merged**